go/dpagg/mean.go

```diff
@@ -213,6 +215,101 @@ func (bm *BoundedMeanFloat64) Result() float64 {
+	if !bm.resultReturned {
+		return noise.ConfidenceInterval{}, fmt.Errorf("Result() must be called before calling ComputeConfidenceInterval()")
+	}
+	// Brute force optimization for finding the tightest confidence interval: Iterate
+	// through a fixed numSplits number of splits for countAlpha where it takes the
+	// range of values strictly from 0 to alpha.
+	increment := alpha / float64(numSplits)
+	tightest := math.Inf(1)
+	var tightestConfInt noise.ConfidenceInterval
+	for curCountAlpha := increment; curCountAlpha < alpha; curCountAlpha += increment {
+		confInt, err := bm.computeConfidenceIntervalSplit(alpha, curCountAlpha)
+		if err != nil {
+			return noise.ConfidenceInterval{}, err
+		}
+		tightness := confInt.UpperBound - confInt.LowerBound
+		if tightness < tightest {
+			tightest = tightness
+			tightestConfInt = confInt
+		}
+	}
+	return tightestConfInt, nil
+}
+
+// computeConfidenceIntervalSplit computes a confidence interval for mean with a given 1 - countAlpha
+// confidence level for count. It computes the corresponding alpha for sum to return a confidence interval
+// with a confidence level of at least 1 - meanAlpha. The computation is based exclusively on the confidence
+// intervals for sum and count. Thus no privacy budget is consumed by this operation.
+//
+// Result() needs to be called before ComputeConfidenceInterval, otherwise this will return an error.
+func (bm *BoundedMeanFloat64) computeConfidenceIntervalSplit(meanAlpha, countAlpha float64) (noise.ConfidenceInterval, error) {
+	if !bm.resultReturned {
+		return noise.ConfidenceInterval{}, fmt.Errorf("Result() must be called before calling computeConfidenceIntervalSplit()")
+	}
+	if err := checkSplit(meanAlpha, countAlpha); err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+
+	sumAlpha := (meanAlpha - countAlpha) / (1 - countAlpha)
+	sumConfInt, err := bm.normalizedSum.ComputeConfidenceInterval(sumAlpha)
+	if err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+	countConfInt, err := bm.count.ComputeConfidenceInterval(countAlpha)
+	if err != nil {
+		return noise.ConfidenceInterval{}, err
+	}
+
+	// BoundedMean maximizes the denominator (count) to 1.
+	countConfInt.LowerBound = math.Max(countConfInt.LowerBound, 1)
+	countConfInt.UpperBound = math.Max(countConfInt.UpperBound, 1)
+
+	// Estimate the lower and upper bounds as minimum and maximum values for mean using lower and
+	// upper bounds of confidence intervals for sum and count.
+	var meanLowerBound, meanUpperBound float64
+	if sumConfInt.LowerBound >= 0 {
+		meanLowerBound = sumConfInt.LowerBound / countConfInt.UpperBound
+	} else {
+		meanLowerBound = sumConfInt.LowerBound / countConfInt.LowerBound
+	}
+	if sumConfInt.UpperBound >= 0 {
+		meanUpperBound = sumConfInt.UpperBound / countConfInt.LowerBound
+	} else {
+		meanUpperBound = sumConfInt.UpperBound / countConfInt.UpperBound
+	}
+	// We are using normalizedSum so we need to add midPoint to both bounds.
```

Comment:
can you maybe refer to the original formulas or at least algorithm at the Result() method?